src/ff/prime_field.rs

```diff
@@ -8,7 +8,13 @@ macro_rules! field_impl {
             type Output = Self;
 
             fn add(self, rhs: Self) -> Self::Output {
-                Self((self.0 + rhs.0) % Self::PRIME)
+                let (result, did_overflow) = self.0.overflowing_add(rhs.0);
```

Comment:
this looks dubious - can we just use regular overflow check? given that we branch here anyway

```rust
fn modulo_add(a: u32, b: u32) -> u32 {
    if a < PRIME - b {
        // no overflow
        a + b
    } else {
        // overflow, wrap around PRIME
        let d = PRIME - b;
        a - d
    }
}
```

Reply:
That logic is incorrect. In the `else` case you're returning
```
a + PRIME - b
```
when what we want is:
```
a + b - PRIME
```

I think the code I have *is* what we want. The problem is that we just have to handle integer overflow when you add two very large u32 values. We can fortify this with a bunch of tests (I already have a few that helped me catch issues).

If we want to eliminate the branch, we can do so like this:
```
let (result, did_overflow) = self.0.overflowing_add(rhs.0);
Self((result + u32::from(did_overflow) * Self::PRIME_TO_INTMAX_DELTA) % Self::PRIME)
```

Reply:
@martinthomson - how would you suggest we implement this?

Reply:
Promoting to the next larger integer type might be easier.

Self::Integer::try_from((Self::OverflowInt::from(a) + Self::OverflowInt::from(b)) % Self::OverflowInt::from(Self::PRIME)).unwrap()

Reply:
Both approaches seem rather wasteful. In the one case, that's a lot of extra copy stuff around. In the other, we are losing another bit of statistical security.

Reply:
Upconversion to u64 is free in practice as long as you are in a register (I believe).

Reply:
OK, free sounds good =). I'll go with that.

Reply:
agree with Martin that upcasting to u64 is free, however I am not sure if we can make a blanket implementation for every field that does that. It can be perfectly fine and optimal for 32 bit field, but for 64 bit and/or default implementation I'd rather go with manual overflow check - apart from my buggy code, I am not sure I understand what is the issue there

Reply:
This would work for additions, but don't we have the same issue for multiplications? right now everything gets upcasted to `u128` to multply, I wonder if it is going to offset all the gains we get by using `u64` in additions